From a4681b15551578c23aa8f569a2a005c73a250ff3 Mon Sep 17 00:00:00 2001
From: Manikanta Pubbisetty <mpubbise@codeaurora.org>
Date: Tue, 7 Mar 2023 10:33:59 +0800
Subject: [PATCH] mac80211: strictly check mesh flags for invalid mesh packets

Adding stricter check based on reserved bits in mesh control flags
to discard invalid mesh packets.

Change-Id: I23f23869059ac53d3b57706c68967feaabcc2119
Signed-off-by: Manikanta Pubbisetty <mpubbise@codeaurora.org>
Signed-off-by: Ronghua Zeng <quic_ronghuaz@quicinc.com>
---
 include/linux/ieee80211.h | 2 ++
 net/mac80211/rx.c         | 3 +++
 2 files changed, 5 insertions(+)

diff --git a/include/linux/ieee80211.h b/include/linux/ieee80211.h
index d394c4a..3fa34f2 100644
--- a/include/linux/ieee80211.h
+++ b/include/linux/ieee80211.h
@@ -784,6 +784,8 @@ struct ieee80211s_hdr {
 #define MESH_FLAGS_AE		0x3
 #define MESH_FLAGS_PS_DEEP	0x4
 
+#define MESH_CTRL_FLAGS_MAXVAL	0x2
+
 /**
  * enum ieee80211_preq_flags - mesh PREQ element flags
  *
diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index b72682e..8be0bb0 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -3032,6 +3032,9 @@ ieee80211_rx_h_mesh_fwding(struct ieee80211_rx_data *rx)
 	if (!mesh_hdr->ttl)
 		return RX_DROP_MONITOR;
 
+	if (mesh_hdr->flags > MESH_CTRL_FLAGS_MAXVAL)
+		return RX_DROP_MONITOR;
+
 	if (mesh_hdr->flags & MESH_FLAGS_AE) {
 		struct mesh_path *mppath;
 		char *proxied_addr;
-- 
2.7.4

